sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusMessageRenewOperatorTest.java

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.messaging.servicebus;
+
+import com.azure.messaging.servicebus.implementation.LockContainer;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.test.StepVerifier;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Unit test for {@link ServiceBusMessageRenewOperator}.
+ */
+public class ServiceBusMessageRenewOperatorTest {
+
+    private static final UUID LOCK_TOKEN_UUID = UUID.randomUUID();
+    private static final String LOCK_TOKEN = LOCK_TOKEN_UUID.toString();
+    private static final Duration MAX_AUTO_LOCK_RENEW_DURATION = Duration.ofSeconds(10);
+
+    @Mock
+    LockContainer<LockRenewalOperation> messageLockContainer;
+    @Mock
+    private Function<String, Mono<OffsetDateTime>> renewalFunction;
+
+    @BeforeAll
+    static void beforeAll() {
+        StepVerifier.setDefaultTimeout(Duration.ofSeconds(30));
+    }
+
+    @AfterAll
+    static void afterAll() {
+        StepVerifier.resetDefaultTimeout();
+    }
+
+    @BeforeEach
+    void setup() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @AfterEach
+    void teardown() {
+        Mockito.framework().clearInlineMocks();
+    }
+
+    /**
+     * Test that the function to renew lock is invoked.
+     */
+    @Test
+    void lockRenewed() {
+        // Arrange
+        final int atLeast = 1;
+        final int processingTimeSeconds = 3;
+        final OffsetDateTime lockedUntil = OffsetDateTime.now().plusSeconds(3);
+        final OffsetDateTime renewLockedUntil = lockedUntil.plusSeconds(3);
+        final ServiceBusReceivedMessage message = new ServiceBusReceivedMessage("data".getBytes());
+        message.setLockToken(LOCK_TOKEN_UUID);
+        message.setLockedUntil(lockedUntil);
+
+        when(renewalFunction.apply(LOCK_TOKEN))
+            .thenReturn(Mono.fromCallable(() -> renewLockedUntil));
+
+        final Flux<? extends ServiceBusReceivedMessage> messageSource = Flux.fromArray(new ServiceBusReceivedMessage[]{message});
```

Comment:
Use TestPublisher because we should test multiple renewals.

We should also test that cancellation from downstream.

Reply:
I think we need more tests about the behaviour of this. has many more cases: https://github.com/reactor/reactor-core/blob/master/reactor-core/src/test/java/reactor/core/publisher/FluxMapTest.java

Reply:
added more test